File: 프로그래머스/3/468376. 최고 속도/최고 속도.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <map>
#include <set>
#include <algorithm>

#define INF 2000000000

using namespace std;
using Pos = pair<int, int>;

vector<int> solution(vector<vector<int>> city, vector<vector<int>> road) {
    // 1. 각 지점의 제한 속도 저장 및 가로/세로 도로 분류
    map<Pos, int> limits;
    vector<vector<Pos>> road_points(road.size());
    
    vector<vector<int>> H_roads; // {도로 인덱스, y, x1, x2}
    vector<vector<int>> V_roads; // {도로 인덱스, x, y1, y2}
    
    for (int i = 0; i < road.size(); i++) {
        int x1 = min(road[i][0], road[i][2]);
        int y1 = min(road[i][1], road[i][3]);
        int x2 = max(road[i][0], road[i][2]);
        int y2 = max(road[i][1], road[i][3]);
        int limit = road[i][4];
        
        // 양 끝점 추가
        road_points[i].push_back({x1, y1});
        road_points[i].push_back({x2, y2});
        
        // 카메라 (정중앙) 위치 추가 및 속도 갱신 (더 낮은 속도 우선)
        Pos mid = {(x1 + x2) / 2, (y1 + y2) / 2};
        road_points[i].push_back(mid);
        
        if (limits.find(mid) == limits.end()) limits[mid] = limit;
        else limits[mid] = min(limits[mid], limit);
        
        if (y1 == y2) H_roads.push_back({i, y1, x1, x2});
        else V_roads.push_back({i, x1, y1, y2});
    }
    
    // 2. 도시 좌표를 속해있는 도로의 정점에 추가
    for (auto& c : city) {
        int cx = c[0], cy = c[1];
        for (auto& hr : H_roads) {
            if (cy == hr[1] && hr[2] <= cx && cx <= hr[3]) {
                road_points[hr[0]].push_back({cx, cy});
            }
        }
        for (auto& vr : V_roads) {
            if (cx == vr[1] && vr[2] <= cy && cy <= vr[3]) {
                road_points[vr[0]].push_back({cx, cy});
            }
        }
    }
    
    // 3. 가로 도로와 세로 도로가 만나는 "교차점" 찾아 정점에 추가
    for (auto& hr : H_roads) {
        for (auto& vr : V_roads) {
            if (hr[2] <= vr[1] && vr[1] <= hr[3] && vr[2] <= hr[1] && hr[1] <= vr[3]) {
                road_points[hr[0]].push_back({vr[1], hr[1]});
                road_points[vr[0]].push_back({vr[1], hr[1]});
            }
        }
    }
    
    // 4. 각 도로 내부의 정점들을 정렬하여 인접한 점들끼리 간선(그래프) 연결
    map<Pos, vector<Pos>> graph;
    for (int i = 0; i < road.size(); i++) {
        auto& pts = road_points[i];
        sort(pts.begin(), pts.end());
        // 중복 좌표 제거
        pts.erase(unique(pts.begin(), pts.end()), pts.end());
        
        for (int j = 0; j < (int)pts.size() - 1; j++) {
            graph[pts[j]].push_back(pts[j+1]);
            graph[pts[j+1]].push_back(pts[j]);
        }
    }
    
    // 5. 올바른 다익스트라 수행 (최대 병목 경로 탐색)
    map<Pos, int> maxLimit;
    priority_queue<pair<int, Pos>> pq;
    
    Pos startPos = {city[0][0], city[0][1]};
    pq.push({INF, startPos});
    maxLimit[startPos] = INF;
    
    while (!pq.empty()) {
        int curLimit = pq.top().first;
        Pos u = pq.top().second;
        pq.pop();
        
        // 이미 큐에 있는 값보다 더 나은 경로로 방문된 적 있다면 스킵
        if (maxLimit[u] > curLimit) continue;
        
        for (Pos v : graph[u]) {
            int nextLimit = curLimit;
            if (limits.find(v) != limits.end()) {
                nextLimit = min(nextLimit, limits[v]);
            }
            
            // 더 높은 속도로 도달할 수 있을 때만 갱신 및 큐에 삽입
            if (maxLimit.find(v) == maxLimit.end() || nextLimit > maxLimit[v]) {
                maxLimit[v] = nextLimit;
                pq.push({nextLimit, v});
            }
        }
    }
    
    // 6. 결과 정리
    vector<int> answer;
    for (int i = 1; i < city.size(); i++) {
        Pos target = {city[i][0], city[i][1]};
        int res = maxLimit.count(target) ? maxLimit[target] : 0;
        if (res == INF) res = 0; // 카메라를 한 번도 거치지 않은 경우
        answer.push_back(res);
    }
    
    return answer;
}

File: 프로그래머스/3/468376. 최고 속도/최고 속도.py
import heapq
from collections import defaultdict

def solution(city, road):
    INF = 10**15
    limits = defaultdict(lambda: INF)
    
    H_roads = []
    V_roads = []
    # 각 도로(인덱스)에 속하는 핵심 지점(노드)들을 담을 집합
    road_points = [set() for _ in range(len(road))]
    
    # 1. 도로 양 끝점과 카메라(중앙점) 위치 추가 및 가로/세로 도로 분류
    for i, (x1, y1, x2, y2, lim) in enumerate(road):
        x1, x2 = min(x1, x2), max(x1, x2)
        y1, y2 = min(y1, y2), max(y1, y2)
        
        road_points[i].add((x1, y1))
        road_points[i].add((x2, y2))
        
        mid = ((x1 + x2) // 2, (y1 + y2) // 2)
        road_points[i].add(mid)
        # 같은 지점에 카메라가 여러 개일 경우 최소 제한 속도를 따름
        limits[mid] = min(limits[mid], lim)
        
        if y1 == y2:
            H_roads.append((i, y1, x1, x2))
        else:
            V_roads.append((i, x1, y1, y2))
            
    # 2. 도시 위치를 해당 도로에 추가
    for cx, cy in city:
        for i, y_h, x_h_min, x_h_max in H_roads:
            if cy == y_h and x_h_min <= cx <= x_h_max:
                road_points[i].add((cx, cy))
        for i, x_v, y_v_min, y_v_max in V_roads:
            if cx == x_v and y_v_min <= cy <= y_v_max:
                road_points[i].add((cx, cy))
                
    # 3. 가로-세로 도로의 교차점 찾기
    for i, y_h, x_h_min, x_h_max in H_roads:
        for j, x_v, y_v_min, y_v_max in V_roads:
            if x_h_min <= x_v <= x_h_max and y_v_min <= y_h <= y_v_max:
                road_points[i].add((x_v, y_h))
                road_points[j].add((x_v, y_h))
                
    # 4. 그래프 생성 (각 도로의 점들을 정렬하여 인접한 점끼리 간선 연결)
    graph = defaultdict(list)
    for i in range(len(road)):
        pts = sorted(list(road_points[i]))
        for k in range(len(pts) - 1):
            u, v = pts[k], pts[k+1]
            graph[u].append(v)
            graph[v].append(u)
            
    # 5. 다익스트라 (Maximum Bottleneck Path) 탐색
    start = tuple(city[0])
    max_speed = defaultdict(int)
    max_speed[start] = INF
    
    # 최대힙을 위해 음수 사용: (-유지 가능한 최고 속도, 현재 노드)
    pq = [(-INF, start)]
    
    while pq:
        curr_speed_neg, u = heapq.heappop(pq)
        curr_speed = -curr_speed_neg
        
        # 이미 더 나은 속도로 도달한 적이 있다면 스킵
        if curr_speed < max_speed[u]:
            continue
            
        for v in graph[u]:
            v_limit = limits[v]
            # 다음 노드로 이동할 때의 속도: 현재까지의 속도와 다음 위치 카메라 제한 속도 중 작은 값
            nxt_speed = min(curr_speed, v_limit)
            
            if nxt_speed > max_speed[v]:
                max_speed[v] = nxt_speed
                heapq.heappush(pq, (-nxt_speed, v))
                
    # 6. 결과 도출 (2번 도시부터 n번 도시까지)
    ans = []
    for k in range(1, len(city)):
        c = tuple(city[k])
        speed = max_speed[c]
        # 제한 속도 카메라를 한 번도 안 만난 경우 (무한대) 0 반환
        if speed == INF:
            ans.append(0)
        else:
            ans.append(speed)
            
    return ans